oeps/oep-0008.rst

```diff
@@ -0,0 +1,323 @@
+
+
+Task Management
+---------------
+We use Makefiles for task management. The reason for this choice is that
+almost any system no matter the OS should be able to use them without any
+special software installation or configuration.
+
+If the Makefile gets too complex, use it as a wrapper to `Paver`_ scripts,
+which are written in Python.
+
+Standard Makefile Targets:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
++--------------+--------------------------------------------------------------+
+| target       | description                                                  |
++==============+==============================================================+
+| help         | display a help message                                       |
++--------------+--------------------------------------------------------------+
+| pip-compile  | update the ``requirements/*.txt`` files with the             |
+|              | latest packages satisfying ``requirements/*.in``             |
++--------------+--------------------------------------------------------------+
+| requirements | install development environment requirements                 |
++--------------+--------------------------------------------------------------+
+| quality      | check coding style                                           |
++--------------+--------------------------------------------------------------+
+| test         | run all unit, integration, and acceptance tests for the repo |
++--------------+--------------------------------------------------------------+
+| validate     | run tests and quality checks                                 |
++--------------+--------------------------------------------------------------+
+| e2e          | run End to End tests against a deployed environment to       |
+|              | verify the system as a user, initiating their actions        |
+|              | through the UI                                               |
++--------------+--------------------------------------------------------------+
+
+.. _Travis CI: https://travis-ci.org/
+.. _PyPI: https://pypi.python.org/pypi
+.. _Paver: https://pythonhosted.org/Paver/
+
+
+Python Language Specific
+------------------------
+
+Requirements
+^^^^^^^^^^^^
+
+* Core requirements are listed in the ``install_requires`` section of
+  ``setup.py``, and only modified when code changes are made that require a
+  new package or drop support for a previously supported one.
+
+* Requirements are managed via `pip-tools`_ and organized by which ones need
+  to be installed for different Python virtual environments.
+
+* Requirements files are kept either at root level or within a directory named
+  ``requirements``.
+
+* Direct dependencies are listed in ``requirements/*.in`` for use in
+  development and `Tox`_ virtual environments with as few version constraints
+  as possible.
+
+* Full pinned dependencies in ``requirements/*.txt`` are auto-generated via
+  ``make pip-compile``.
```

Comment:
For private packages, we can use a `private.txt` file, and have pip-sync install that, but have it `.gitignore`d so that it doesn't get checked in.

For pinning: we can explicitly pin things in the `.in` files, and `pip-compile` will respect that. `--upgrade` will call for upgrading everything, but we don't have to actually check in the changes to the `.txt` files in one chunk. It will also be easier to manage if we can do better about staying on top of dependency changes.

@jzoldak: There are differences in our strategies for libraries vs IDAs (in particular, libraries should use requirements files for their development requirements, but IDAs use them for production requirements as well). Do we need to call those differences out here?